### PR TITLE
Judge v1beta1

### PR DESCRIPTION
**Pull request by gearnode**
**Goal:** Increase test coverage of apiserver package to refactor this package safely.

Close #32

### CODE REVIEW

**Comment by gearnode** [comment]
I should refactor how test apiserver to have readable and maintainable test suite.

**Ideas:**
- Use factory like to build apiserver request more easy.
- Create custom matcher to ensure apiserver response.

**Comment by gearnode** [comment]
I change the scope of this pull request by a global refactor of the application and target a v1beta1 with this.

File: pkg/apiserver/getpolicy.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+	"google.golang.org/grpc/status"
+)
+
+// GetPolicy implement judge.api.v1alpha1.Judge.GetPolicy
+func (s *Service) GetPolicy(ctx context.Context, in *v1alpha1.GetPolicyRequest) (*v1alpha1.Policy, error) {
+
+	pol, err := s.state.GetPolicy(in.GetOrn())
+	if err != nil {
+		log.Error(err)
+		return &v1alpha1.Policy{}, status.Error(codes.NotFound, "the policy with the orn "+in.GetOrn()+" is not found")
```
Comment (lukkor) [inline pkg/apiserver/getpolicy.go]:
```suggestion
		return &v1alpha1.Policy{}, status.Error(codes.NotFound, "the policy with the orn %q is not found", in.GetOrn())
```

File: pkg/storage/memory/memory_store.go [hunk 2]
```diff
@@ -14,100 +14,69 @@ See the License for the specific language governing permissions and
...
-	i := 0
-	for _, v := range partition {
-		data[i] = v
-		i++
+	if s.policies[id] != nil {
+		return s.policies[id], nil
 	}
 
-	return data, nil
+	return nil, errors.New("policy does not exist")
```
Comment (lukkor) [inline pkg/storage/memory/memory_store.go]:
```suggestion
	return nil, errors.Errorf("policy with id %q does not exist", id)
```

File: pkg/storage/memory/memory_store.go [hunk 3]
```diff
@@ -14,100 +14,69 @@ See the License for the specific language governing permissions and
...
 }
 
-// DescribeAll return all element in the given table
-func (store *MemoryStore) DescribeAll(table string) ([]interface{}, error) {
-	store.mux.RLock()
-	partition := store.data[table]
-	store.mux.RUnlock()
-	data := make([]interface{}, len(partition))
+// GetPolicy returns the policy with the given identifier.
+func (s *MemoryStore) GetPolicy(id string) (*policy.Policy, error) {
```
Comment (lukkor) [inline pkg/storage/memory/memory_store.go]: You should switch between error case and normal case IMO.

File: pkg/storage/memory/memory_store.go [hunk 4]
```diff
@@ -14,100 +14,69 @@ See the License for the specific language governing permissions and
...
 
-	if elem == nil {
-		return nil, errors.New("record not exist")
+	if s.policies[id] != nil {
+		delete(s.policies, id)
+		return nil
 	}
 
-	return elem, nil
+	return errors.New("policy does not exist")
```
Comment (lukkor) [inline pkg/storage/memory/memory_store.go]:
```suggestion
	return nil, errors.Errorf("policy with id %q does not exist", id)
```

File: pkg/authorize/authorize_test.go [hunk 5]
```diff
@@ -1,14 +1,28 @@
...
 import (
-	"testing"
-
 	"github.com/gearnode/judge/pkg/authorize"
 	"github.com/gearnode/judge/pkg/orn"
 	"github.com/gearnode/judge/pkg/policy"
 	"github.com/gearnode/judge/pkg/policy/resource"
-
 	"github.com/stretchr/testify/assert"
+	"testing"
```
Comment (lukkor) [inline pkg/authorize/authorize_test.go]: The idiom is to put standard package in a section at the top of the list.

File: pkg/authorize/authorize.go [hunk 6]
```diff
@@ -46,43 +44,61 @@ var (
 // The order in which the policies are evaluated has no effect on the outcome
 // of the evaluation. All policies are evaluated, and the result is always
 // that the request is either allowed or denied.
-func Authorize(policies []*policy.Policy, who orn.ORN, what string, something orn.ORN, context map[string]string) (bool, error) {
-	// TODO: remove storage dependencies
+func Authorize(policies []*policy.Policy, _subject orn.ORN, action string, object orn.ORN, _context map[string]string) (bool, error) {
```
Comment (lukkor) [inline pkg/authorize/authorize.go]: You should precise the purpose of each argument as this method looks important.

Comment (lukkor) [reply pkg/authorize/authorize.go]: I added this comment as I do not understand `_context` purpose.

**Pull request merged**